sdk/src/assertions/collection_hash.rs

```diff
@@ -0,0 +1,288 @@
+use std::{
+    fs::File,
+    io::{Read, Seek},
+    path::{Component, Path, PathBuf},
+};
+
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    assertions::AssetType, asset_handlers::zip_io, hash_stream_by_alg,
+    hash_utils::verify_stream_by_alg, Error, HashRange, Result,
+};
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Default)]
+pub struct CollectionHash {
+    pub uris: Vec<UriHashedDataMap>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub alg: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none", with = "serde_bytes")]
+    pub zip_central_directory_hash: Option<Vec<u8>>,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct UriHashedDataMap {
+    pub uri: PathBuf,
+
+    #[serde(with = "serde_bytes")]
+    pub hash: Vec<u8>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub size: Option<u64>,
+
+    #[serde(rename = "dc:format", skip_serializing_if = "Option::is_none")]
+    pub dc_format: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub data_types: Option<Vec<AssetType>>,
+}
+
+impl CollectionHash {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn add_uri_map(&mut self, uri_map: UriHashedDataMap) {
+        self.uris.push(uri_map);
+    }
+
+    // The base path MUST be the folder of the manifest. A URI MUST NOT reference a path outside of that folder.
+    pub fn gen_hash<R>(&mut self, base_path: &Path) -> Result<()>
```

Comment:
I think base_path should be part of the constructor and this function should just calc the hashes when requested.

Reply:
The problem is that base_path is only used when the assertion is for non-zips. We can have a constructor that takes in a path, something like with_base_path, but then we'd have to store it optionally. If the assertion is for a non-zip and constructed without a base path, then it's already invalid.

Reply:
Thinking about it again, we should only have public methods to construct a collection hash with a base path. When using collection hash for zips, user's should never need to construct a collection hash explicitly.